src/platforms/esp32/components/avm_sys/platform_nifs.c

```diff
@@ -628,6 +629,82 @@ static term nif_crypto_hash(Context *ctx, int argc, term argv[])
     return term_from_literal_binary(digest, digest_len, &ctx->heap, ctx->global);
 }
 
+static const AtomStringIntPair cipher_table[] = {
+    { ATOM_STR("\xB", "aes_128_cbc"), MBEDTLS_CIPHER_AES_128_CBC },
+    { ATOM_STR("\xB", "aes_128_ctr"), MBEDTLS_CIPHER_AES_128_CTR },
+    SELECT_INT_DEFAULT(MBEDTLS_CIPHER_NONE)
+};
+
+static term nif_crypto_crypto_one_time(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+
+    term cipher_term = argv[0];
+    mbedtls_cipher_type_t cipher
+        = interop_atom_term_select_int(cipher_table, cipher_term, ctx->global);
+    if (UNLIKELY(cipher == MBEDTLS_CIPHER_NONE)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    term key = argv[1];
+    VALIDATE_VALUE(key, term_is_binary);
+    term iv = argv[2];
+    VALIDATE_VALUE(iv, term_is_binary);
+    term data = argv[3];
+    VALIDATE_VALUE(data, term_is_binary);
+    term flag_or_options = argv[4];
+    mbedtls_operation_t operation;
+    switch (flag_or_options) {
+        case TRUE_ATOM:
+            operation = MBEDTLS_ENCRYPT;
+            break;
+        case FALSE_ATOM:
+            operation = MBEDTLS_DECRYPT;
+            break;
+        default:
+            // TODO: add option list support
+            RAISE_ERROR(BADARG_ATOM);
+    }
+
+    const unsigned char *key_data = (const unsigned char *) term_binary_data(key);
+    size_t key_len_bits = term_binary_size(key) * 8;
+
+    const mbedtls_cipher_info_t *cipher_info = mbedtls_cipher_info_from_type(cipher);
+
+    mbedtls_cipher_context_t cipher_ctx;
+    mbedtls_cipher_setup(&cipher_ctx, cipher_info);
+    mbedtls_cipher_setkey(&cipher_ctx, key_data, key_len_bits, operation);
+
+    unsigned int block_size = mbedtls_cipher_get_block_size(&cipher_ctx);
+
+    size_t data_size = term_binary_size(data);
+    size_t needed_mem = term_binary_heap_size(data_size);
+    if (UNLIKELY(memory_ensure_free(ctx, needed_mem) != MEMORY_GC_OK)) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+
+    iv = argv[2];
+    const unsigned char *iv_data = (const unsigned char *) term_binary_data(iv);
+    int iv_len = term_binary_size(iv);
+    data = argv[3];
+    const unsigned char *data_data = (const unsigned char *) term_binary_data(data);
+
+    size_t temp_buf_size = data_size + block_size;
+    void *temp_buf = malloc(temp_buf_size);
```

Comment:
Why not avoiding allocating twice and calling term_create_uninitialized_binary directly here?

Reply:
` mbedtls_cipher_crypt` requires as an output buffer `data_size` + `block_size`. But only `data_size` bytes are relevant, the others are just used internally from ` mbedtls_cipher_crypt`.
The other option for avoiding buffer copies would be creating a binary that is never returned and returning a sub-binary to the buffer visible portion.

Reply:
I think the semantic is wrong here. This is where tests would be handy.

```erlang
Key = <<1:128>>,
IV = <<0:128>>,
Txt = <<"First bytesSecond bytes">>,
<<231,108,83,104,188,131,182,65,2,128,78,162,210,149,128,248>> = crypto:crypto_one_time(aes_128_cbc, Key, IV, Txt, true).
```

Reply:
Removed auto padding.